src/device/pf_cmrpc_helpers.c

```diff
@@ -53,3 +98,309 @@ void pf_generate_uuid (
+   p_handle->handle_uuid.time_low = (timestamp & 0x0000FFFF);
+   p_handle->handle_uuid.time_mid = (timestamp & 0x00FF0000);
+   p_handle->handle_uuid.time_hi_and_version = (timestamp & 0xFF000000) |
+                                               0x1000;
+   p_handle->handle_uuid.clock_hi_and_reserved = 0x80;
+   p_handle->handle_uuid.clock_low = /* 0x11 */ 0x0C;
+   p_handle->handle_uuid.node[0] = mac_address.addr[0];
+   p_handle->handle_uuid.node[1] = mac_address.addr[1];
+   p_handle->handle_uuid.node[2] = mac_address.addr[2];
+   p_handle->handle_uuid.node[3] = mac_address.addr[3];
+   p_handle->handle_uuid.node[4] = mac_address.addr[4];
+   p_handle->handle_uuid.node[5] = mac_address.addr[5];
+}
+
+/**
+ * Initialize protocol tower configuration including floors
+ * @param p_tower    Out: Protocol tower configuration
+ * @param epm_type   In: End point type, EPM or PNIO
+ * @param udp_port   In: Upd protocol port
+ */
+static void pf_init_rpc_tower_entry (
+   pf_rpc_tower_t * p_tower,
+   pf_rpc_epm_type_t epm_type,
+   uint16_t udp_port)
+{
+   /* Floor 1 EPMv4 */
+   p_tower->floor_1.protocol_id = PF_RPC_PROTOCOL_UUID;
+   if (epm_type == PF_EPM_TYPE_EPMV4)
+   {
+      memcpy (
+         &p_tower->floor_1.uuid,
+         &uuid_epmap_interface,
+         sizeof (p_tower->floor_1.uuid));
+      p_tower->floor_1.version_major = RPC_FLOOR_VERSION_EPMv4;
+   }
+   else
+   {
+      memcpy (
+         &p_tower->floor_1.uuid,
+         &uuid_io_device_interface,
+         sizeof (p_tower->floor_1.uuid));
+      p_tower->floor_1.version_major = RPC_FLOOR_VERSION_PNIO;
+   }
+   p_tower->floor_1.version_minor = RPC_FLOOR_VERSION_MINOR;
+
+   /* Floor 2 32bit NDR*/
+   p_tower->floor_2.protocol_id = PF_RPC_PROTOCOL_UUID;
+   memcpy (&p_tower->floor_2.uuid, &uuid_ndr, sizeof (p_tower->floor_2.uuid));
+   p_tower->floor_2.version_major = RPC_FLOOR_VERSION_NPR;
+   p_tower->floor_2.version_minor = RPC_FLOOR_VERSION_MINOR;
+
+   /* Floor 3 RPC*/
+   p_tower->floor_3.protocol_id = PF_RPC_PROTOCOL_CONNECTIONLESS;
+   p_tower->floor_3.version_minor = RPC_FLOOR_VERSION_MINOR;
+
+   /* Floor 4 UDP Listening port*/
+   p_tower->floor_4.protocol_id = PF_RPC_PROTOCOL_DOD_UDP;
+   p_tower->floor_4.port = htons (udp_port);
+
+   /* Floor 5 IP Address */
+   p_tower->floor_5.protocol_id = PF_RPC_PROTOCOL_DOD_IP;
+   p_tower->floor_5.ip_address = 0;
+}
+
+/**
+ * Add EPMv4 endpoint to lookup response
+ * @param net              InOut: The p-net stack instance
+ * @param p_rpc_req        In:    Rpc header.
+ * @param p_lookup_req     In:    Lookup request.
+ * @param p_lookup_rsp     Out:   Lookup response.
+ * @param p_read_status    Out:   Read pnio status
+ * @param res_size         In:    The size of the output buffer.
+ * @param p_res            Out:   The output buffer.
+ * @param p_pos            InOut: Position in the output buffer.
+ * @return  0  if operation succeeded.
+ *          -1 if an error occurred.
+ */
+static int pf_cmrdr_add_epmv4_entry (
```

Comment:
The documentation does not seem to match. Please make input pointer arguments const.